Topics/Development_Process/Error_Handling_in_ExpressJS.md

```diff
@@ -0,0 +1,184 @@
+4. [Custom Error Objects](#custom-error-objects)
+    - [Status Codes](#status-codes)
+    - [Implementing a Custom Error class](#implementing-a-custom-error-class)
+    - [Example Usage](#example-usage)
+5. [Conclusion](#conclusion)
+
+## Introduction
+Express.JS (aka Express) is a scalable backend web framework for building RESTful APIs in the Node.JS programming language. It provides a variety of interelated features for server-side API development, including support for routing, middleware, templating (e.g. with EJS and Handlebars), static file serving and utilities for managing HTTP request-response flows. Error handling is an important aspect of any sofware project as it is critical for development, maintainability and functionality. It is especially important when building RESTful APIs where there exists specific expectations of how erroneous scenarios should be handled, such as setting appropriate status codes in response objects, returning error information in a consistent format etc.
+
+This guide aims to highlight the mechanisms in Express that afford graceful error handling and how they can be utilised with good practices. This article will assume knowledge of Express and that an operational Express server is set up with some API endpoints that carry out some logic.
+
+## Goals of Error Handling for RESTful APIs
+* **Appropriate status codes are set:** It is important that the HTTP status code is appropriate and as specific as possible with regard to the error that has occurred. Information about what each HTTP status code represents can be found [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). For example, if a GET request for a resource fails because there are no resources corresponding to a provided identifying parameter, a 404 *"Not Found"* status code should be returned. Poor error handling will often be reflected by all error responses setting a  500 *"Internal Server Error* (typically used as generic catch-all status code but should be only set as a last resort).
+* **Informative Error Message:** The response body for an erroneous request must include an error message that is insightful, concise and highlighting strictly relevant information (i.e. it should not be populated with garbage). This information will vary on a case-by-case basis, but relevant variables should generally be featured (such as an ID for a resource involved in the thrown error) while avoiding including too many variables such that the message is not readable. Custom error classes may be designed to associate additional metadata with a request without overwhelming the message string.
+* **Robust Logging:** Every erroneous request should be logged appropriately. Graceful error handling should ensure that if a developer receives an error response for a request, they should be able to trace the request within the application's log, with at least one entry corresponding to the request error.
```

Comment:
It would be a good idea to provide specific recommendations or examples for more advanced logging libraries.

Reply:
I see your point, however, I do not believe that that is in the scope of this article and may unnecessarily complicate things. I think that showing simply the placement of a logger call is sufficient for the scope of the guide.

Reply:
After looking into the topic of logging, that is fair, it may overly complicate this article.